Tipp service 96

VolkerHBZ opened the pull request.

VolkerHBZ committed:
deprecating coverage properties that are now represended in TIPPCoverageStatement

VolkerHBZ committed:
removed call of the not-implemented method KBComponent.addToTags

VolkerHBZ committed:
ComponentUpdateService.ensureCoreData can handle data represented as JSON and (new!) KBComponent

VolkerHBZ committed:
TippService and Test

VolkerHBZ committed:
handle find conflicts

VolkerHBZ committed:
copy name and identifiers from the TitleInstance to connected TIPPs

VolkerHBZ committed:
moved determineTitleClass from IntegrationController to TitleInstance

VolkerHBZ committed:
ConcurrencyManager.Job for TIPP copy title data

VolkerHBZ committed:
added hibernate session flushes and making enabled the job to be canceled

VolkerHBZ committed:
added hibernate session flushes and making enabled the job to be canceled

hornmo commented on server/gokbg3/grails-app/services/org/gokb/TippService.groovy:
```diff
@@ -0,0 +1,128 @@
+    else if (found.to_create == true) {
+      ti = Class.forName(title_class_name).newInstance()
+      ti.name = tipp.name
+      ti.ids = tipp.ids
```
The new title has to be saved (and flushed) before doing anything with Combos.

hornmo commented on server/gokbg3/grails-app/services/org/gokb/TippService.groovy:
```diff
@@ -0,0 +1,128 @@
+      ti.ids = tipp.ids
+    }
+    // Add the core data.
+    if (ti) {
```
If we are doing this with matched titles as well as new titles we may overwrite existing title info here, which is not necessarily desired. Restricting this to new titles for now seems better. We may want to work with ReviewRequests for changed title data instead.

VolkerHBZ replied:
I supposed that copying the way the IntegrationController works, by using the ComponentpdateService methods, it'd give a similarly desired result including the ReviewRequests and won't puzzle eventual users using the new mechanism more than before.

VolkerHBZ replied:
done: if the title is found in the database, it's not changed except for being linked to the tipp.

hornmo commented on server/gokbg3/grails-app/services/org/gokb/TippService.groovy:
```diff
@@ -0,0 +1,128 @@
+                                                               firstAuthor     : tipp.firstAuthor,
+                                                               firstEditor     : tipp.firstEditor]))
+      }
+
```
Since we will have a publisherName in the TIPP, we should also try to set this on the title as well (via titleLookupService.addPublisher)

VolkerHBZ replied:
maybe implementing a copy constructor in TitleInstance and it's subclasses?

VolkerHBZ committed:
added hibernate session flushes and making enabled the job to be canceled

VolkerHBZ committed:
changed appearances of <TIPP>.title to <TIPP>.title? in appropriate places

VolkerHBZ committed:
testing matchPackage()

VolkerHBZ committed:
new Integration endpoint updatePackageTipps

VolkerHBZ committed:
making updatePackageTipps test work.

VolkerHBZ committed:
code beautifying

VolkerHBZ committed:
removed "editionNumber" property
updatePackageTipps endpoint for importing tipps without title matching
added Property TIPP.importId for KBART title_id data

VolkerHBZ committed:
working matching of tipps by their identifiers

VolkerHBZ committed:
prevent multiple identical RefdataValues

VolkerHBZ committed:
Zwischenstand

VolkerHBZ committed:
Zwischenstand

VolkerHBZ committed:
titlematching job works

VolkerHBZ committed:
essearchservice.find() defunct.

hornmo committed:
fix ESSearch handling

hornmo committed:
Merge pull request #444 from openlibraryenvironment/tippService_96_fix_esfind

fix ESSearch handling

VolkerHBZ committed:
further development

VolkerHBZ committed:
fixing tests

VolkerHBZ committed:
Merge remote-tracking branch 'origin/g3_master' into tippService_96

# Conflicts:
#	server/gokbg3/grails-app/services/com/k_int/ESSearchService.groovy

VolkerHBZ committed:
UpdatePackageRun testing

VolkerHBZ committed:
fixing NonUniqueDomainObjectException by NOT reusing the stale RefdataValue Combo.Status=Active from a former Hibernate Session.

VolkerHBZ committed:
Merge branch 'tippService_96' of https://github.com/openlibraryenvironment/gokb into tippService_96

 Conflicts:
	server/gokbg3/src/main/groovy/org/gokb/UpdatePkgTippsRun.groovy

VolkerHBZ committed:
new test case

VolkerHBZ committed:
new test case

VolkerHBZ committed:
new test case

VolkerHBZ committed:
new test case

VolkerHBZ committed:
Merge remote-tracking branch 'origin/g3_master' into tippService_96

# Conflicts:
#	server/gokbg3/grails-app/services/org/gokb/TippService.groovy
#	server/gokbg3/src/main/groovy/org/gokb/DomainClassExtender.groovy

VolkerHBZ committed:
code cleanup

VolkerHBZ committed:
finishing touches

VolkerHBZ committed:
Review Requests

VolkerHBZ committed:
added title matching after tipp import

VolkerHBZ committed:
session handling

VolkerHBZ committed:
session handling

VolkerHBZ committed:
more logging

VolkerHBZ committed:
revert session handling

VolkerHBZ committed:
handling ti.publisher == null

VolkerHBZ committed:
handling publicationType == null

VolkerHBZ committed:
handling publicationType == null
adapting tests

VolkerHBZ committed:
fix tests

VolkerHBZ committed:
inject AutoTimestampEventListener

VolkerHBZ committed:
fixing tests

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1224,7 +1295,7 @@ class IntegrationController {
       }
       else {
-        def title_class_name = determineTitleClass(titleObj)
+        def title_class_name = TitleInstance.determineTitleClass(titleObj.publitionType?:titleObj.type)
```
`publitionType` -> `publicationType`

VolkerHBZ replied:
done

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/IntegrationController.groovy:
```diff
@@ -1299,7 +1370,7 @@ class IntegrationController {
 
             if (!result.message) {
-              result.message = messageService.resolveCode('crossRef.title.success', null, locale)
+              result.message = messageService.resolveCode('crossRef.title.success', [title], locale)
```
Message parameter should be `title.name`

VolkerHBZ replied:
ok

hornmo commented on server/gokbg3/grails-app/controllers/org/gokb/PublicController.groovy:
```diff
@@ -185,12 +185,12 @@ class PublicController {
-                            sanitize( tipp.title.getCurrentPublisher()?.name ) + '\t' +
-                            sanitize( tipp.title.getPrecedingTitleId() ) + '\t' +
+                            sanitize( tipp.title?.getCurrentPublisher()?.name ) + '\t' +
+                            sanitize( tipp.title?.getPrecedingTitleId() ) + '\t' +
```
We should fallback to TIPP info if no title info is available.

VolkerHBZ replied:
ok

hornmo commented on server/gokbg3/grails-app/domain/org/gokb/cred/TitleInstancePackagePlatform.groovy:
```diff
@@ -976,4 +1089,16 @@ class TitleInstancePackagePlatform extends KBComponent {
     result
   }
+
+  public static RefdataValue determinePubTypeRef(String someType) {
+    if (someType) {
+      RefdataValue[] pubTypes = RefdataCategory.lookup(TitleInstancePackagePlatform.RD_PUBLICATION_TYPE)
+      for (RefdataValue pubType : pubTypes) {
+        if (someType.equalsIgnoreCase(pubType.value)) {
+          return pubType
+        }
+      }
+    }
+    return null
+  }
```
Why not just do `RefdataCategory.lookup(TitleInstancePackagePlatform.RD_PUBLICATION_TYPE, someType)`? No need to iterate manually then.

VolkerHBZ replied:
This was implemented to circumvent problems with case-sensitive matching (in the rdv_cache)

I'll try your suggestion.

VolkerHBZ replied:
works

hornmo commented on server/gokbg3/grails-app/services/org/gokb/FTUpdateService.groovy:
```diff
@@ -103,6 +104,10 @@ class FTUpdateService {
         result.updater = 'org'
+        result.titleNamespace = kbc.titleNamespace?.value
+        result.packageNamespace = kbc.packageNamespace?.value
+        result.updater = 'org'
+        result.updater = 'org'
```
Remove, duplicate lines to 106

VolkerHBZ replied:
ok.

hornmo commented on server/gokbg3/grails-app/services/org/gokb/TippService.groovy:
```diff
@@ -1,41 +1,125 @@
+        matchTitle(tipp)
+      }
+    }
+  }
```
We can get around looking up every TIPP by default by specifying the conditions (combo to package, no title combo) in an hql and only getting a list of IDs of affected TIPPs.

hornmo commented on server/gokbg3/grails-app/services/org/gokb/TippService.groovy:
```diff
@@ -1,41 +1,125 @@
+    } else if (found.to_create == true) {
+      ti = Class.forName(title_class_name).newInstance()
+      ti.name = tipp.name
+      titleLookupService.addPublisher(tipp.publisherName, ti)
```
New component has to be saved before adding Combos. Also this is called in line 78 again, so this may be removed.

hornmo commented on server/gokbg3/grails-app/services/org/gokb/TippService.groovy:
```diff
@@ -73,4 +157,33 @@ class TippService {
+          "TIPP ${tipp.name} coudn't be linked.",
+          null,
+          null,
+          found.matches as JSON,
```
Since `matches` returned from TitleLookupService currently contains full components in `object`, these have to be explicitly mapped to JSON. Existing ReviewRequests hold all other linked components as maps in an array `additionalInfo.otherComponents` of the form `[oid: <obj.logEntityId>, name: <obj.name>]`.

VolkerHBZ replied:
Mapping the matched components to that format

hornmo commented on server/gokbg3/grails-app/services/org/gokb/TippService.groovy:
```diff
@@ -73,4 +157,33 @@ class TippService {
+          "TIPP ${tipp.name} conflicts with other titles.",
+          null,
+          null,
+          found.conflicts as JSON,
```
See above

hornmo commented on server/gokbg3/src/main/groovy/org/gokb/UpdatePkgTippsRun.groovy:
```diff
@@ -0,0 +1,730 @@
+      }
+
+      pkg = Package.get(proxy.id)
+      def dummy = pkg.provider
+      dummy = pkg.nominalPlatform
```
Unused?

VolkerHBZ replied:
deleted

VolkerHBZ committed:
using determinePubType Method

hornmo commented on server/gokbg3/src/main/groovy/org/gokb/UpdatePkgTippsRun.groovy:
```diff
@@ -0,0 +1,730 @@
+      TypeConvertingMap map = [
+          componentType    : 'TitleInstancePackagePlatform',
+          importId         : tippJson.titleId,
+          tippPackageUuid  : pkg.uuid,
+          hostPlatformUuid : tippJson.hostPlatform.uuid,
```
Use `pkg` & `platform` as query params for `find()`, as documented at https://github.com/openlibraryenvironment/gokb/wiki/Elasticsearch-API (component-specific fields).

VolkerHBZ committed:
requested changes

VolkerHBZ committed:
completed HQL for finding TIPPs without Title

VolkerHBZ committed:
completed HQL for finding TIPPs without Title

VolkerHBZ committed:
completed HQL for finding TIPPs without Title

VolkerHBZ committed:
renamed parameters for ES find() query

VolkerHBZ committed:
Merge remote-tracking branch 'origin/g3_master' into tippService_96

# Conflicts:
#	server/gokbg3/grails-app/domain/org/gokb/cred/TitleInstancePackagePlatform.groovy
#	server/gokbg3/grails-app/services/org/gokb/TippService.groovy

VolkerHBZ committed:
merge with g3_master

VolkerHBZ committed:
merge with g3_master

hornmo committed:
user check & test cleanup
